### PR TITLE
std.crypto.gimli enhancements

### PR DESCRIPTION
**Pull request by daurnimator**
- Add `std.mem.secureEqual`
  - Add code for AEAD mode
  - Have crypto hash benchmark include gimli

### CODE REVIEW

File: lib/std/mem.zig [hunk 1]
```diff
@@ -365,6 +365,32 @@ pub fn eql(comptime T: type, a: []const T, b: []const T) bool {
     return true;
 }
 
+/// Compares two slices in constant time and returns whether they are equal.
+pub fn secureEqual(comptime T: type, a: []const T, b: []const T) bool {
```
[the remark this answers is not in the document]
Comment (daurnimator) [reply lib/std/mem.zig]: I just copied naming convention from `secureZero`.

File: lib/std/mem.zig [hunk 2]
```diff
@@ -365,6 +365,32 @@ pub fn eql(comptime T: type, a: []const T, b: []const T) bool {
...
+    if (a.len != b.len) return false;
+    if (a.ptr == b.ptr) return true;
```
[the remark this answers is not in the document]
Comment (daurnimator) [reply lib/std/mem.zig]: The algorithm just needs to be constant time in terms of not exiting at the first non-matching byte. Exiting early on equivalent pointers should be fine. 
Alternatively, we could assert that `a.ptr != b.ptr` and make that a requirement of the API.

File: lib/std/mem.zig (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewrk) [reply lib/std/mem.zig]: But `secureZero` is unrelated to constant time.

File: lib/std/mem.zig (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewrk) [reply lib/std/mem.zig]: Even without the early exits here, it's still not sound. The function destroys the information obtained with the xor. It is possible for the optimizer to detect this and exit early.

[the remark this answers is not in the document]
Comment (daurnimator) [reply lib/std/mem.zig]: Okay... so.... is the correct answer to turn off all optimizations for this function?

[the remark this answers is not in the document]
Comment (andrewrk) [reply lib/std/mem.zig]: That doesn't really solve the problem. Zig (as well as most other Ahead-Of-Time compiled languages) is currently fundamentally incapable of modeling constant-time algorithms.

Until #1776 is solved, don't even try to address the use case. Use `std.mem.eql` and document that the implementation is vulnerable to timing attacks.

There are still plenty of use cases for using Gimli and other cryptographic implementations with non-constant-time implementations. For example, offline batch processing. This patch is still useful, but we can't claim that it solves timing attacks. The best course forward for now is to document that preventing timing attacks are currently out of scope for this implementation.

**Pull request merged**